crypto/x509/x509aset.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/err.h>
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+#include "x509_acert.h"
+
+static int replace_gentime(ASN1_STRING **dest, const ASN1_GENERALIZEDTIME *src)
+{
+    ASN1_STRING *s;
+
+    if (src->type != V_ASN1_GENERALIZEDTIME)
+        return 0;
+
+    if (*dest == src)
+        return 1;
+
+    s = ASN1_STRING_dup(src);
+    if (s == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_ASN1_LIB);
+        return 0;
+    }
+
+    ASN1_STRING_free(*dest);
+    *dest = s;
+
+    return 1;
+}
+
+static int replace_dirName(GENERAL_NAMES **names, const X509_NAME *dirName)
+{
+    GENERAL_NAME *gen_name = NULL;
+    STACK_OF(GENERAL_NAME) *new_names = NULL;
+    X509_NAME *name_copy;
+
+    if ((name_copy = X509_NAME_dup(dirName)) == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_ASN1_LIB);
+        goto err;
+    }
+
+    if ((new_names = sk_GENERAL_NAME_new_null()) == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_ASN1_LIB);
+        goto err;
+    }
+
+    if ((gen_name = GENERAL_NAME_new()) == NULL) {
+        ERR_raise(ERR_LIB_X509, ERR_R_ASN1_LIB);
+        goto err;
+    }
+
+    if (sk_GENERAL_NAME_push(new_names, gen_name) <= 0) {
+        ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);
```

Comment:
You may factorise the code in this method:
- only call `ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);` after end label
- it will spare some accolades.
- none of these point of failures are ASN1 related.

Reply:
`GENERAL_NAME_new` is defined by the `IMPLEMENT_ASN1_FUNCTIONS`.  Doesn't that make its failure ASN1 related?  That seems consistent with how the errors are handled in [several](https://github.com/openssl/openssl/blob/master/crypto/x509/v3_akid.c#L208) [other](https://github.com/openssl/openssl/blob/master/crypto/ess/ess_lib.c#L97) [places](https://github.com/openssl/openssl/blob/master/crypto/x509/v3_san.c#L452).

Maybe I'm not understanding something fundamental about these ERR reason codes.  I thought that they should explain the reason that the caller exited abnormally.  So if the caller calls an ASN1 related function and that function fails, ASN1 should be the reason code (unless there is a more specific reason).  Is that not the case?

> only call ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB); after end label

If different failure conditions are raising different errors, then calling ERR_raise in each if block seems the correct thing to do.  If  `ERR_R_CRYPTO_LIB` is the correct code for all error cases in this function it can be refactored as you suggest.  That will depend on the resolution of the ASN1 error code question.

Reply:
> - only call `ERR_raise(ERR_LIB_X509, ERR_R_CRYPTO_LIB);` after end label

Er, why?  The source line number is registered at the `ERR_raise()` point, so conglomerating them to one call after `end:` will make tracing harder.

Reply:
Another good point.  

Are there any objections to resolving this comment, and leaving the code as-is?

Reply:
Resolving